src/cycle_basis.c

```diff
@@ -0,0 +1,660 @@
+ * if that edge is present in the graph.
+ *
+ * \param graph The input graph, it might be directed, but edge
+ *    direction is ignored.
+ * \param basis A pointer to a vector of vectors. Each fundamental cycle
+ *    will be a member as a list of edge ids.
+ * \return Error code.
+ * 
+ * Time complexity: O(|V| + |E|) (spanning tree) + ?? (backtracking).
+ * 
+ */
+int igraph_fundamental_cycle_basis(const igraph_t *graph,
+    igraph_vector_ptr_t *basis) {
+
+    long int no_of_nodes=igraph_vcount(graph);
+    long int no_of_edges=igraph_ecount(graph);
+    long int no_of_cycles=no_of_edges - (no_of_nodes - 1);
+    igraph_vector_char_t added_nodes, added_edges;
+    igraph_vector_t parent_nodes, parent_edges, dist, inc_edges;
+    igraph_dqueue_t q=IGRAPH_DQUEUE_NULL;
+
+    long int i, j, k;
+
+    // Given a spanning tree (forest), fundamental cycles are given by
+    // unions of:
+    // - a path within the tree
+    // - a single edge connecting end to start nodes outside the tree
+    // The number of fundamental cycles is #edges - (#vertices -1),
+    // because the quantity in parenthesis is the number of edges in a
+    // tree (no cycles).
+    
+    // The key is that every non-tree edge generates a fundamental cycle
+    // so we run a BFS and whenever we encounter a vertex which is not
+    // new we are:
+    // - either trying to walk back the last edge we just walked
+    // - or using a non-tree edge (since both vertices are known already)
+    // Note: every edge must be walked once, so even after the tree
+    // is complete we still have to finish walking the remaining edges.
+    // Note: we cannot use the igraph_bfs function because we have to
+    // inspect outgoing edges that point to vertices that have already
+    // been added
+
+    IGRAPH_VECTOR_CHAR_INIT_FINALLY(&added_nodes, no_of_nodes);
+    IGRAPH_VECTOR_CHAR_INIT_FINALLY(&added_edges, no_of_edges);
+    IGRAPH_VECTOR_INIT_FINALLY(&parent_nodes, no_of_nodes);
+    IGRAPH_VECTOR_INIT_FINALLY(&parent_edges, no_of_edges);
+    IGRAPH_VECTOR_INIT_FINALLY(&dist, no_of_nodes);
+    IGRAPH_DQUEUE_INIT_FINALLY(&q, 100);
+    IGRAPH_VECTOR_INIT_FINALLY(&inc_edges, 0);
+
+    for (i = 0; i < no_of_nodes; i++) {
+        /* Start of a new connected component including the first one */
+        if (added_nodes[i] > 0) { continue; }
+
+        IGRAPH_ALLOW_INTERRUPTION();
+
+	/* This is the root, mark it visited */
+        added_nodes[i] = 1;
+
+	/* The queue stores the next vertex, the edge to it, and its distance to root */
+        IGRAPH_CHECK(igraph_dqueue_push(&q, i));
+        IGRAPH_CHECK(igraph_dqueue_push(&q, -1));
+        IGRAPH_CHECK(igraph_dqueue_push(&q, 0));
+        while (! igraph_dqueue_empty(&q)) {
+          /* current node, incoming edge, and distance from tree root */
+          long int actnode=(long int) igraph_dqueue_pop(&q);
+	  long int actedge=(long int) igraph_dqueue_pop(&q);
```

Comment:
Let's not use tab characters for indentation.

Reply:
Speaking about tabs vs spaces: I think we have an `.editorconfig` at the root, but not all editors pick up the settings from `.editorconfig` by default, so we need another safety mechanism to prevent misformatted code from slipping into the codebase. Initially I wanted to propose using [pre-commit](https://pre-commit.com/) for validating the codebase with pre-commit hooks, but then I realized that third-party contributors would probably not have `pre-commit` installed and then we would be back to square one. Do you guys know about any Github plugin that could be used to validate code style in the PRs, or do we have to create a CI test for that?